File: src_python/qnprj/prjop_4.py
import sys
import numpy as np
import cython

import crsys as crs
import qnnum as qnn
import qnvec as qnv
import qnmat as qnm
import qnmath as qmt
import qnndarray as qna 

# alias for prjop_i
def projection3(v: qnv.Qnvec) -> qnv.Qnvec:
    return prjvec_i(v)

# alias for prjop
def projection_numerical(vn: qnv.Qnvec) -> qnv.Qnvec:
    return prjvec(vn)

def projection3_sets_numerical(vns: qna.QnNdarray) -> qna.QnNdarray:
    #Parameters
    #vsn: qnvector array
    #    set of nD vectors shape (num,n) assumed or
    #    (num,3,n) or (num,4,n) for triangles or tetrahedra 
    shape=vns.shape
    ndim=len(shape)
    #print("shape",shape,"ndim",ndim)  # for test
    isys=crs.isys
    if isys>2:
        ni=2
    else:
        ni=3
    if ndim==1:
        m=qna.zeros((ni,))
        m=projection3(vns)
    elif ndim==2:
        m=qna.zeros((shape[0],ni))
        for i in range(shape[0]):
            m[i]=projection3(vns[i])
    elif ndim==3:
        m=qna.zeros((shape[0],shape[1],ni))
        for i in range(shape[0]):
            for j in range(shape[1]):
                m[i][j]=projection3(vns[i][j])
    return m

# alias for prjop_e
def projection_numerical_par(v: qnv.Qnvec) -> qnv.Qnvec:
    return prjvec_e(v)

# alias for prjop_i
def get_internal_component_numerical(v: qnv.Qnvec) -> qnv.Qnvec:
    return prjvec_i(v)

def check_ltv(n,N):
    # check lattice vector external and internal space components
    ndv=np.ndarray(3**3,dtype=qnn.Qnnum)
    n=5
    V0=qnn.any(([ 0, 0, 1]))
    for i1 in range(-1,2):
        for i2 in range(-1,2):
            for i3 in range(-1,2):
                for i4 in range(-1,2):
                    V1=qnn.any([ i1, 0, 1])
                    V2=qnn.any([ i2, 0, 1])
                    V3=qnn.any([ i3, 0, 1])
                    V4=qnn.any([ i4, 0, 1])
                    #V0=qnn.any(np.array([ 0, 0, 1]),N)
                    #VT=np.array([v1,v2,v3,V4,V0])
                    vt=qnv.anyv(n,N,[V1,V2,V3,V4,V0])
                    qnn.printqnv("ndv",vt) #print qnvector expression
                    qnv=prj0@vt  #qnv=vt@prj0  #@ndv #external enternal components
                    qnn.printqnv("qnv",qnv) #print qnvector expression
                    n+=1
                    
def print_prj(str:str,prj:qnm.Qnmat):
    #print(prj)
    print(str)
    n=prj.shape[0]
    N=prj.N
    print("n",n,"N",N)
    
    qnm.printqnm(str,prj)
    
    b=qnm.copy(prj)
    print(str)
    qnm.printqnm("str",b)
    
def printfm(str,prj3f,n):
    print(str)
    for i in range(n):
        for j in range(n):
            #print(prj3f[i][j],end=" ")
            print("{:8f}".format(prj3f[i][j]),end=" ")
        print()
    print()
        
def qnm2flnm(prj:qnm.Qnmat):
    n=prj.shape[0]
    prj0f=np.ndarray((n,n),dtype=float)
    for i in range(n):
        for j in range(n):
            a=prj[i][j]
            #print("a",a.n[0],a.n[1],a.n[2])  # for test
            prj0f[i][j]=qnn.qn2flt(a)*scl
            #print("f",prj0f[i][j])  # for test
    return prj0f

File: src_python/dihed/qnprj/prjop_4.cpp
#include <iostream>
#include <vector>
#include <array>
#include <cmath>
#include "crsys.h"
#include "qnnum.h"
#include "qnvec.h"
#include "qnmat.h"
#include "qnmath.h"
#include "qnndarray.h"

using namespace std;

// alias for prjop_i
Qnvec projection3(Qnvec v) {
    return prjvec_i(v);
}

// alias for prjop
Qnvec projection_numerical(Qnvec vn) {
    return prjvec(vn);
}

QnNdarray projection3_sets_numerical(QnNdarray vns) {
    // Parameters
    // vsn: qnvector array
    // set of nD vectors shape (num,n) assumed or
    // (num,3,n) or (num,4,n) for triangles or tetrahedra 
    auto shape = vns.shape();
    int ndim = shape.size();
    int isys = crs::isys;
    int ni = (isys > 2) ? 2 : 3;
    QnNdarray m;

    if (ndim == 1) {
        m = QnNdarray(ni);
        m = projection3(vns);
    } else if (ndim == 2) {
        m = QnNdarray(shape[0], ni);
        for (int i = 0; i < shape[0]; i++) {
            m[i] = projection3(vns[i]);
        }
    } else if (ndim == 3) {
        m = QnNdarray(shape[0], shape[1], ni);
        for (int i = 0; i < shape[0]; i++) {
            for (int j = 0; j < shape[1]; j++) {
                m[i][j] = projection3(vns[i][j]);
            }
        }
    }
    return m;
}

// alias for prjop_e
Qnvec projection_numerical_par(Qnvec v) {
    return prjvec_e(v);
}

// alias for prjop_i
Qnvec get_internal_component_numerical(Qnvec v) {
    return prjvec_i(v);
}

void check_ltv(int n, int N) {
    // check lattice vector external and internal space components
    vector<Qnnum> ndv(27); // 3^3
    n = 5;
    Qnnum V0 = Qnnum({0, 0, 1});
    for (int i1 = -1; i1 <= 1; i1++) {
        for (int i2 = -1; i2 <= 1; i2++) {
            for (int i3 = -1; i3 <= 1; i3++) {
                for (int i4 = -1; i4 <= 1; i4++) {
                    Qnnum V1 = Qnnum({i1, 0, 1});
                    Qnnum V2 = Qnnum({i2, 0, 1});
                    Qnnum V3 = Qnnum({i3, 0, 1});
                    Qnnum V4 = Qnnum({i4, 0, 1});
                    // V0 = Qnnum(array({0, 0, 1}), N);
                    // VT = array({v1, v2, v3, V4, V0});
                    auto vt = Qnvec::anyv(n, N, {V1, V2, V3, V4, V0});
                    qnn::printqnv("ndv", vt); // print qnvector expression
                    auto qnv = prj0 * vt; // qnv = vt * prj0; // external internal components
                    qnn::printqnv("qnv", qnv); // print qnvector expression
                    n++;
                }
            }
        }
    }
}

void print_prj(const string& str, Qnmat prj) {
    // print(prj);
    cout << str << endl;
    int n = prj.shape()[0];
    int N = prj.N;
    cout << "n: " << n << ", N: " << N << endl;

    qnm::printqnm(str, prj);

    Qnmat b = qnm::copy(prj);
    cout << str << endl;
    qnm::printqnm("str", b);
}

void printfm(const string& str, const Qnmat& prj3f, int n) {
    cout << str << endl;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cout << fixed << setprecision(8) << prj3f[i][j] << " ";
        }
        cout << endl;
    }
    cout << endl;
}

vector<vector<double>> qnm2flnm(Qnmat prj) {
    int n = prj.shape()[0];
    vector<vector<double>> prj0f(n, vector<double>(n));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            auto a = prj[i][j];
            // cout << "a: " << a.n[0] << ", " << a.n[1] << ", " << a.n[2] << endl; // for test
            prj0f[i][j] = qnn::qn2flt(a) * scl;
            // cout << "f: " << prj0f[i][j] << endl; // for test
        }
    }
    return prj0f;
}

